### PR TITLE
🌱 migrate kcp clientset

### PR DESCRIPTION
**Pull request by stevekuznetsov**
Depends on https://github.com/kcp-dev/kubernetes/pull/111

Error: failed: Your PR description is *really* short.
Error: failed: Your PR description is *really* short.
Error: failed: Your PR description is *really* short.
Error: failed: Your PR description is *really* short.
Error: failed: Your PR description is *really* short.
Error: failed: Your PR description is *really* short.
Error: failed: Your PR description is *really* short.
Error: failed: Your PR description is *really* short.
Error: failed: Your PR description is *really* short.
Error: failed: Your PR description is *really* short.
Error: failed: Your PR description is *really* short.

### CODE REVIEW

File: pkg/admission/clusterworkspacetypeexists/admission_test.go [hunk 1]
```diff
@@ -63,10 +63,10 @@ func updateAttr(obj, old *tenancyv1alpha1.ClusterWorkspace) admission.Attributes
 	return admission.NewAttributesRecord(
 		helpers.ToUnstructuredOrDie(obj),
 		helpers.ToUnstructuredOrDie(old),
-		tenancyv1alpha1.Kind("ClusterWorkspace").WithVersion("v1alpha1"),
+		tenancyv1alpha1.Kind("ClusterWorkspace").WithVersion("tenancyv1alpha1listers"),
```
Comment (sttts) [inline pkg/admission/clusterworkspacetypeexists/admission_test.go:66]: this is wrong

Comment (sttts) [reply pkg/admission/clusterworkspacetypeexists/admission_test.go:66]: many more below

File: pkg/authorization/workspace_content_authorizer_test.go [hunk 2]
```diff
@@ -433,8 +434,12 @@ func TestWorkspaceContentAuthorizer(t *testing.T) {
 				return
 			}
 
-			if got := recordingAuthorizer.recordedAttributes.GetUser(); !reflect.DeepEqual(got, tt.wantUser) {
-				t.Errorf("want user %+v, got %+v", tt.wantUser, got)
+			if recordingAuthorizer.recordedAttributes == nil {
+				t.Errorf("want user %+v, got %+v", tt.wantUser, nil)
+			} else {
```
Comment (sttts) [inline pkg/authorization/workspace_content_authorizer_test.go:439]: `else if got`

File: pkg/proxy/server.go [hunk 3]
```diff
@@ -53,8 +50,7 @@ func NewServer(ctx context.Context, c CompletedConfig) (*Server, error) {
 	s := &Server{
 		CompletedConfig: c,
 	}
-	rootShardConfigInformerConfig := kcpclienthelper.SetCluster(restclient.CopyConfig(s.CompletedConfig.RootShardConfig), tenancyv1alpha1.RootCluster)
-	rootShardConfigInformerClient, err := kcpclient.NewForConfig(rootShardConfigInformerConfig)
+	rootShardConfigInformerClient, err := kcpclientset.NewForConfig(s.CompletedConfig.RootShardConfig)
```
Comment (sttts) [inline pkg/proxy/server.go:53]: @p0lyn0mial please check

**Comment by sttts** [comment]
/lgtm
/approve

Comments can be addressed in follow-up. Let's avoid rebase-hell.

**Review by fgiloux — COMMENTED** [review]
A few non critical questions

File: pkg/proxy/server.go (hunk 3 above)
Comment (p0lyn0mial) [reply pkg/proxy/server.go:53]: How does it know about the `root` cluster? I think it is implicit, it relies on the assumption that an admin will set a valid kubeconfig context.

I'd prefer to explicitly set it to the `root` cluster and then to use `NewSharedScopedInformerFactory`, wdyt?

File: pkg/reconciler/workload/placement/placement_controller.go [hunk 4]
```diff
@@ -63,19 +63,17 @@ func NewController(
...
 	}
 
-	indexers.AddIfNotPresentOrDie(
-		c.placementIndexer,
-		cache.Indexers{
-			indexers.PlacementByLocationWorkspace: indexers.IndexPlacementByLocationWorkspace,
-		},
-	)
+	if err := placementInformer.Informer().AddIndexers(cache.Indexers{
+		byLocationWorkspace: indexByLocationWorkspace,
```
[the remark this answers is not in the document]
Comment (stevekuznetsov) [reply pkg/reconciler/workload/placement/placement_controller.go:73]: I have no energy to fix the 3000000 merge conflicts from that PR. I will re-do it once this merges instead.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -105,7 +105,7 @@ If you find yourself working on something that is unplanned and/or untracked (i.
     - `pclusterClient`
     - `rootKcpClient`
     - `orgKubeClient`
-- Configs intended for `NewClusterForConfig` (i.e. today often called "admin workspace config") should uniformly be called `clusterConfig`
+- Configs intended for `NewForConfig` (i.e. today often called "admin workspace config") should uniformly be called `clusterConfig`
```
[the remark this answers is not in the document]
Comment (stevekuznetsov) [reply CONTRIBUTING.md:108]: I think the alternative is `clusterConfig` and `clusterClient` for cluster-aware, unscoped. `config` and `client` (as in k8s) for cluster-specific. WDYT?

File: pkg/proxy/server.go (hunk 3 above)
Comment (stevekuznetsov) [reply pkg/proxy/server.go:53]: If we want a cluster-scoped LIST+WATCH, yes, 100% you are correct @p0

File: CONTRIBUTING.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (fgiloux) [reply CONTRIBUTING.md:108]: I am fine with that. Could we have it worded in a similar way in CONTRIBUTING.md? People, who have used kcp may understand "cluster-aware" and "unscoped" to a specific workspace but what is "intended for NewForConfig" may not be obvious to them, at least it was not to me.

**Comment by stevekuznetsov** [comment]
Forgot to bump client-go as well ... re-tagging

**Pull request merged**